File: levanta/ios/Levanta/Nucleo/Contador.swift
import Foundation

/// Os exercícios que desligam o alarme.
enum Exercicio: String, CaseIterable, Codable {
    case agachamento, flexao, polichinelo, abdominal

    var rotulo: String {
        switch self {
        case .agachamento: "Agachamento"
        case .flexao: "Flexão de braço"
        case .polichinelo: "Polichinelo"
        case .abdominal: "Abdominal"
        }
    }

    var instrucao: String {
        switch self {
        case .agachamento: "Desça até a coxa ficar paralela ao chão e volte a ficar em pé."
        case .flexao: "Desça até o cotovelo fechar e empurre até esticar o braço."
        case .polichinelo: "Abra braços e pernas ao mesmo tempo e feche de volta."
        case .abdominal: "Suba até o tronco chegar perto dos joelhos e volte a deitar."
        }
    }

    var enquadramento: String {
        switch self {
        case .agachamento: "Apoie o celular no chão ou numa cadeira, a uns 2 metros, com o corpo inteiro na tela."
        case .flexao: "Deite o celular de lado, no chão, a uns 2 metros, filmando você de perfil."
        case .polichinelo: "Celular em pé a uns 2,5 metros — precisa caber você de braços abertos."
        case .abdominal: "Celular de lado, no chão, filmando você de perfil com os joelhos dobrados."
        }
    }

    var repeticoesSugeridas: Int {
        switch self {
        case .agachamento: 20
        case .flexao: 10
        case .polichinelo: 30
        case .abdominal: 20
        }
    }
}

/// Em que ponto do movimento o corpo está.
enum Fase { case procurando, repouso, extremo }

/// O que a tela precisa dizer para a pessoa acertar o movimento.
enum Aviso: String {
    case corpoForaDeQuadro = "Afaste o celular até aparecer o corpo inteiro"
    case posicaoInvalida = "Ajuste a posição para o exercício escolhido"
    case descaMais = "Desça mais"
    case subaMais = "Volte à posição inicial"
    case maisDevagar = "Mais devagar, o movimento não contou"

    var mensagem: String { rawValue }
}

/// O resultado de processar um quadro.
struct Progresso {
    let repeticoes: Int
    let fase: Fase
    /// 0 no repouso, 1 no extremo — alimenta a barra de amplitude da tela.
    let amplitude: Float
    let aviso: Aviso?
    /// `true` só no quadro exato em que a repetição fechou.
    let contouAgora: Bool
}

/// Conta repetições a partir de um sinal escalar com histerese.
///
/// Todo exercício vira um único número que é **alto em repouso e baixo no
/// extremo** do movimento. Uma repetição é o ciclo `repouso → extremo →
/// repouso`. Os dois limiares separados impedem meia repetição de contar, e os
/// tempos mínimos impedem uma sacudida do celular de virar agachamento.
///
/// Esta é a tradução literal de `Contador.kt`: mesmos limiares, mesma
/// suavização, mesmos tempos. As duas versões são exercitadas pelos mesmos
/// casos de teste para não divergirem com o tempo.
class ContadorRepeticoes {
    let exercicio: Exercicio

    init(exercicio: Exercicio) { self.exercicio = exercicio }

    // Pontos de extensão das subclasses.
    func sinal(_ e: Esqueleto) -> Float? { nil }
    func postura(_ e: Esqueleto) -> Aviso? { nil }
    var limiarRepouso: Float { 0 }
    var limiarExtremo: Float { 0 }
    var suavizacao: Float { 0.55 }
    var msNoExtremo: Int64 { 180 }
    var msPorRepeticao: Int64 { 500 }

    private(set) var repeticoes = 0
    private var fase: Fase = .procurando
    private var suave: Float?
    private var entrouNoExtremoMs: Int64 = 0
    private var saiuDoRepousoMs: Int64 = 0
    private var avisoPendente: Aviso?

    func reiniciar() {
        repeticoes = 0
        fase = .procurando
        suave = nil
        entrouNoExtremoMs = 0
        saiuDoRepousoMs = 0
        avisoPendente = nil
    }

    func processar(_ e: Esqueleto) -> Progresso {
        guard let bruto = sinal(e) else {
            suave = nil
            return Progresso(repeticoes: repeticoes, fase: .procurando, amplitude: 0,
                             aviso: .corpoForaDeQuadro, contouAgora: false)
        }
        if let erro = postura(e) {
            suave = nil
            return Progresso(repeticoes: repeticoes, fase: .procurando, amplitude: 0,
                             aviso: erro, contouAgora: false)
        }

        let v = suave.map { $0 + suavizacao * (bruto - $0) } ?? bruto
        suave = v

        let amplitude = 1 - faixa(v, limiarExtremo, limiarRepouso)
        var contou = false

        switch fase {
        case .procurando:
            // Só entra no ciclo pelo repouso: quem começa a filmar já agachado
            // espera ficar em pé para a primeira repetição valer inteira.
            if v >= limiarRepouso {
                fase = .repouso
                avisoPendente = nil
            } else {
                avisoPendente = .subaMais
            }

        case .repouso:
            if v <= limiarExtremo {
                fase = .extremo
                entrouNoExtremoMs = e.instanteMs
                if saiuDoRepousoMs == 0 { saiuDoRepousoMs = e.instanteMs }
                avisoPendente = nil
            } else if v < limiarRepouso {
                if saiuDoRepousoMs == 0 { saiuDoRepousoMs = e.instanteMs }
                avisoPendente = .descaMais
            } else {
                saiuDoRepousoMs = 0
                avisoPendente = nil
            }

        case .extremo:
            if v >= limiarRepouso {
                let tempoNoExtremo = e.instanteMs - entrouNoExtremoMs
                let tempoDoCiclo = e.instanteMs - saiuDoRepousoMs
                if tempoNoExtremo >= msNoExtremo && tempoDoCiclo >= msPorRepeticao {
                    repeticoes += 1
                    contou = true
                    avisoPendente = nil
                } else {
                    avisoPendente = .maisDevagar
                }
                fase = .repouso
                saiuDoRepousoMs = 0
            } else {
                avisoPendente = nil
            }
        }

        return Progresso(repeticoes: repeticoes, fase: fase, amplitude: amplitude,
                         aviso: avisoPendente, contouAgora: contou)
    }

    /// Escolhe entre esquerda e direita o lado mais visível, e mede o sinal nele.
    func melhorLado(_ e: Esqueleto, _ medir: (Esqueleto, Lado) -> Float?) -> Float? {
        let esq = medir(e, .esquerdo)
        let dir = medir(e, .direito)
        guard let esq else { return dir }
        guard let dir else { return esq }
        // Os dois lados visíveis: a média cancela o ruído de um marco tremendo.
        return abs(esq - dir) <= 25 ? (esq + dir) / 2 : min(esq, dir)
    }
}

/// Fábrica: o contador certo para o exercício escolhido.
func contadorDe(_ exercicio: Exercicio) -> ContadorRepeticoes {
    switch exercicio {
    case .agachamento: ContadorAgachamento()
    case .flexao: ContadorFlexao()
    case .polichinelo: ContadorPolichinelo()
    case .abdominal: ContadorAbdominal()
    }
}

File: levanta/android/core/src/main/kotlin/br/com/levanta/core/pose/Contador.kt
package br.com.levanta.core.pose

import kotlin.math.abs

/** Os exercícios que desligam o alarme. */
enum class Exercicio(val rotulo: String, val instrucao: String, val enquadramento: String) {
    AGACHAMENTO(
        rotulo = "Agachamento",
        instrucao = "Desça até a coxa ficar paralela ao chão e volte a ficar em pé.",
        enquadramento = "Apoie o celular no chão ou numa cadeira, a uns 2 metros, com o corpo inteiro na tela.",
    ),
    FLEXAO(
        rotulo = "Flexão de braço",
        instrucao = "Desça até o cotovelo fechar e empurre até esticar o braço.",
        enquadramento = "Deite o celular de lado, no chão, a uns 2 metros, filmando você de perfil.",
    ),
    POLICHINELO(
        rotulo = "Polichinelo",
        instrucao = "Abra braços e pernas ao mesmo tempo e feche de volta.",
        enquadramento = "Celular em pé a uns 2,5 metros — precisa caber você de braços abertos.",
    ),
    ABDOMINAL(
        rotulo = "Abdominal",
        instrucao = "Suba até o tronco chegar perto dos joelhos e volte a deitar.",
        enquadramento = "Celular de lado, no chão, filmando você de perfil com os joelhos dobrados.",
    );

    val repeticoesSugeridas: Int
        get() = when (this) {
            AGACHAMENTO -> 20
            FLEXAO -> 10
            POLICHINELO -> 30
            ABDOMINAL -> 20
        }
}

/** Em que ponto do movimento o corpo está. */
enum class Fase {
    /** Ainda não deu para ler a pose. */
    PROCURANDO,

    /** Posição de partida: em pé, braço esticado, deitado, corpo fechado. */
    REPOUSO,

    /** Ponto extremo do movimento: agachado, peito no chão, tronco em cima, corpo aberto. */
    EXTREMO,
}

/** O que a tela precisa dizer para a pessoa acertar o movimento. */
enum class Aviso(val mensagem: String) {
    CORPO_FORA_DE_QUADRO("Afaste o celular até aparecer o corpo inteiro"),
    POSICAO_INVALIDA("Ajuste a posição para o exercício escolhido"),
    DESCA_MAIS("Desça mais"),
    SUBA_MAIS("Volte à posição inicial"),
    MAIS_DEVAGAR("Mais devagar, o movimento não contou"),
}

/** O resultado de processar um quadro. */
data class Progresso(
    val repeticoes: Int,
    val fase: Fase,
    /** 0 no repouso, 1 no extremo — alimenta a barra de progresso do movimento. */
    val amplitude: Float,
    val aviso: Aviso?,
    /** `true` só no quadro exato em que a repetição fechou. */
    val contouAgora: Boolean,
)

/**
 * Conta repetições a partir de um sinal escalar com histerese.
 *
 * Todo exercício aqui vira um único número que é **alto em repouso e baixo no
 * extremo** do movimento: o ângulo do joelho no agachamento, o do cotovelo na
 * flexão, o do quadril no abdominal, e um índice de fechamento no polichinelo.
 * Uma repetição é o ciclo completo `repouso → extremo → repouso`.
 *
 * Os dois limiares separados são o que impede meia repetição de contar: para
 * sair do repouso o sinal tem que cruzar [limiarExtremo], bem abaixo de
 * [limiarRepouso], então balançar em torno de um único valor não gera contagem.
 * Além disso a repetição só fecha se o extremo tiver durado [msNoExtremo] e o
 * ciclo inteiro [msPorRepeticao] — uma sacudida rápida do celular não vira
 * agachamento.
 */
abstract class ContadorRepeticoes(val exercicio: Exercicio) {

    /** O sinal do quadro, alto em repouso e baixo no extremo. `null` se não der para medir. */
    protected abstract fun sinal(e: Esqueleto): Float?

    /** Postura errada para este exercício (deitado quando devia estar em pé, etc). */
    protected open fun postura(e: Esqueleto): Aviso? = null

    protected abstract val limiarRepouso: Float
    protected abstract val limiarExtremo: Float

    /** Peso do quadro novo na suavização exponencial. Menor = mais estável, mais atrasado. */
    protected open val suavizacao: Float = 0.55f

    /** Tempo mínimo no extremo para o movimento valer. */
    protected open val msNoExtremo: Long = 180

    /** Tempo mínimo do ciclo inteiro. */
    protected open val msPorRepeticao: Long = 500

    private var repeticoes = 0
    private var fase = Fase.PROCURANDO
    private var suave: Float? = null
    private var entrouNoExtremoMs = 0L
    private var saiuDoRepousoMs = 0L
    private var avisoPendente: Aviso? = null

    fun reiniciar() {
        repeticoes = 0
        fase = Fase.PROCURANDO
        suave = null
        entrouNoExtremoMs = 0
        saiuDoRepousoMs = 0
        avisoPendente = null
    }

    fun processar(e: Esqueleto): Progresso {
        val bruto = sinal(e)
        if (bruto == null) {
            suave = null
            return Progresso(repeticoes, Fase.PROCURANDO, 0f, Aviso.CORPO_FORA_DE_QUADRO, false)
        }
        postura(e)?.let { erro ->
            suave = null
            return Progresso(repeticoes, Fase.PROCURANDO, 0f, erro, false)
        }

        val v = suave?.let { it + suavizacao * (bruto - it) } ?: bruto
        suave = v

        val amplitude = 1f - faixa(v, limiarExtremo, limiarRepouso)
        var contou = false

        when (fase) {
            Fase.PROCURANDO -> {
                // Só entra no ciclo pelo repouso: quem começa a filmar já agachado
                // espera ficar em pé para a primeira repetição valer inteira.
                if (v >= limiarRepouso) {
                    fase = Fase.REPOUSO
                    avisoPendente = null
                } else {
                    avisoPendente = Aviso.SUBA_MAIS
                }
            }

            Fase.REPOUSO -> {
                if (v <= limiarExtremo) {
                    fase = Fase.EXTREMO
                    entrouNoExtremoMs = e.instanteMs
                    if (saiuDoRepousoMs == 0L) saiuDoRepousoMs = e.instanteMs
                    avisoPendente = null
                } else if (v < limiarRepouso) {
                    // Saiu do repouso mas ainda não chegou ao extremo.
                    if (saiuDoRepousoMs == 0L) saiuDoRepousoMs = e.instanteMs
                    avisoPendente = Aviso.DESCA_MAIS
                } else {
                    saiuDoRepousoMs = 0L
                    avisoPendente = null
                }
            }

            Fase.EXTREMO -> {
                if (v >= limiarRepouso) {
                    val tempoNoExtremo = e.instanteMs - entrouNoExtremoMs
                    val tempoDoCiclo = e.instanteMs - saiuDoRepousoMs
                    if (tempoNoExtremo >= msNoExtremo && tempoDoCiclo >= msPorRepeticao) {
                        repeticoes++
                        contou = true
                        avisoPendente = null
                    } else {
                        avisoPendente = Aviso.MAIS_DEVAGAR
                    }
                    fase = Fase.REPOUSO
                    saiuDoRepousoMs = 0L
                } else {
                    avisoPendente = null
                }
            }
        }

        return Progresso(repeticoes, fase, amplitude, avisoPendente, contou)
    }

    /** Escolhe entre esquerda e direita o lado mais visível, e mede o sinal nele. */
    protected fun melhorLado(e: Esqueleto, medir: (Esqueleto, Lado) -> Float?): Float? {
        val esq = medir(e, Lado.ESQUERDO)
        val dir = medir(e, Lado.DIREITO)
        if (esq == null) return dir
        if (dir == null) return esq
        // Os dois lados visíveis: a média cancela o ruído de um marco tremendo.
        return if (abs(esq - dir) <= 25f) (esq + dir) / 2f else minOf(esq, dir)
    }
}
